7/25, trending algorithm

samiklein011011 opened the pull request:
Hi Tyler! I created a trending algorithm for events that updates hourly and is based upon the number of votes an event gets, how close the date of the event is to the current date, and the number of users going to the event. It's kind of arbitrary though, so I wanted to see what you think and if you had any recommendations for it. It's in the file "/include/algorithmFunctions.php" and the updating hourly part is in the "include/echoFunctions.php" file, in the echoEvent function around line 270. Thanks!

also: I did a bunch of if statements, because I felt like giving the date difference a weight was more complicated and I think by using if statements it allows for more control over each case (but maybe that's unnecessary?). I was also trying to make there be a big difference between 1 day and 2 days aways, and not very much difference between, for example, 35 and 55 days away.

Commit:
7/13 end of day commit

- working on friend button logic,
- also on the event delete page, so can possibly pop up window that says are you sure you want to do this?

Commit:
7/16

- friending works!! woot. (when you press friend, they become your friend)

- next to do:make it so refreshes friend list in JS right away

Commit:
7/16 friend list refresh in JS,

-refreshes friend list in JS right away (although if you press it twice it doesn't work... i think it has something to do with the DB functions...and that weird console message?)

Commit:
7/18 morning

Commit:
7/18

- fixed intakeFriendship
- made creator of the event click-able
- footer
- trying to make UI softer but also keep it bright, u know

Commit:
7/22 over the weekend, See More button!

- see more button works!!!! (I think!) Very exciting.
- changed the location of some include functions

Commit:
7/23 see more button stuff

- see more button on events the user created
- only shows the see more button if there are more than 3 events

Commit:
7/24 style things, working on trending events algorithm

- fixing style things
- date looks nicer now
- fades in the friends, when you press the friend button
- fade in the upcoming events  when press see all/see less button
- fade in events created when press see all/see less button
- working on trending events algorithm

Commit:
7/25 trending, updates daily

- writing trending algorithm
- updates the trending algorithm daily

Commit:
7/25 hourly recalculates

- recalculates points hourly now
-style things i think

Commit:
7/25 end of the day

- updated algorithm to include users going

tilikang commented on include/algorithmFunctions.php:
```diff
@@ -0,0 +1,107 @@
+	$dateDiff= dateDifference($fullTodayDate, $event['dateOfEvent'])+1;
+
+
+	//NOTE: there is probably a better way to do this than a bunch of if statements,
```
Probably an easier way to do this would be to make an array that maps # of days to a point value. Then you could just loop through that and apply the number of points as soon as you find the appropriate row. That'd save you from having all these `else if` statements.

samiklein011011 replied:
Makes sense. I get how to put a single if equals statement into the array and map it to a point value (like if($dateDiff == 1), give 400 points would be 1->400), but is there anyway to put the ranges into the array too? Like if(15>$dateDiff&&$dateDiff > 5) then map to 100?

tilikang replied:
You could do this in one of two ways. 

The way I wouldn't recommend is to put the start and end range into an array like

```
0=> array(
    'Start'=>5,
    'End'=>15,
    'Points'=>100
)
```

Probably better than that though would be to only have one value, and you assume anything lower than it gets the number of points shown. So you loop through the array until you find an entry with a higher value than the one you're looking for, and give it that number of points.

tilikang commented on include/algorithmFunctions.php:
```diff
@@ -0,0 +1,107 @@
+
+	$points=$points+($event['votes']*1.5);
+	$yMDH=date('Y-m-d-H');
+	updatePoints($event['eventID'], $points, $yMDH);
```
This is a pretty simple algorithm, but sometimes simple is good. If you ever get anyone using this site, you'll probably want to keep an eye on the rankings and make adjustments based on your personal judgement (like, maybe you find that events tomorrow are just as important as events today, or that 5/10/25 aren't the right cut-offs for number of people going). But this seems like a great starting point.

Just to brainstorm ways this could be enhanced in the future:
- You could give extra weight to attendees that are more reputable (have more friends, have attended events in the past, etc.)
- You could personalize the algorithm to show each user different results depending on whether users they're friends with are going to the event. This would require storing a different score for each user
- You could let the hosts of events pay to boost their event's ranking to promote it

I realize you probably won't do any of those things, but it's fun to think about.

tilikang commented on include/echoFunctions.php:
```diff
@@ -114,6 +138,189 @@ function echoDialogBox(){
+function echoEvent($event){
+
+	//re calculate points, hourly
+	$hold = getLastCalculated2($event['eventID']);
```
You already have the whole event here right? Can't you just pull the last calculated date from that?

tilikang commented on include/echoFunctions.php:
```diff
@@ -114,6 +138,189 @@ function echoDialogBox(){
+	//re calculate points, hourly
+	$hold = getLastCalculated2($event['eventID']);
+	$dateNHourCalculated = $hold['0']['lastCalculated2'];
+	if($dateNHourCalculated !==  date('Y-m-d-H')){
```
This is a clever way to handle this. Just so you know, the "normal" way to update these scores periodically would be to run a "cron job" which is basically a tool to run a web page every so often. So you could make a script that updates the score for each event, and run that script every hour.

But it requires some server tools to set up a cron job, so you can't really do that. So this is a pretty cool way to handle it.

tilikang reviewed (commented):
This looks great. I wish I had more constructive comments, but it seems like you've got everything pretty well figured out.

Commit:
7/26 going & save from UI, associative array for trending alg

- made some of the if statements into an associative php array
- created Going from the UI
- can now privately save events - they only show up on your profile & other people can't see them

Commit:
7/26 end of day, lit just fixed buttons

- lit just fixed the buttons lol. but it's the end of the day :)

Commit:
7/27 sum stuff

- see all/less for private events
- edit event button is out of the whitebox
- working on sort By stuff

Commit:
7/27 attempt at sort by

- attempting to do sort by
- not working out too well

Commit:
7/30 working on sort by

- working on sortBy
- categories clickable

Commit:
7/31 end of day

- working on sort by
- changed location of js files to outside the include folder
- color in background of categories in the create event page

Commit:
8/1 end of day, fixed the button on the sorting

- fixing up the sort by page

Commit:
8/2 small bugs

- small bug fixes
